File: lcof/of006/Solution.go
package Solution

type ListNode struct {
	Val  int
	Next *ListNode
}

//	递归写法
func printListFromTailToHead(head *ListNode) []int {
	ans := make([]int, 0)
	if head == nil {
		return ans
	}
	ans = printListFromTailToHead(head.Next)
	ans = append(ans, head.Val)
	return ans
}

//	反转琏表
func printListFromTailToHead2(head *ListNode) []int {
	pre, cur, next, ans := &ListNode{}, head, head.Next, []int{}
	for cur != nil {
		next = cur.Next
		cur.Next = pre

		pre = cur
		cur = next
	}
	for pre.Next != nil {
		ans = append(ans, pre.Val)
		pre = pre.Next
	}
	return ans
}

File: lcof/of006/Solution.py
from typing import List


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        return self.reversePrint(head.next) + [head.val] if head else []


class Solution2:
    def reversePrint(self, head: ListNode) -> List[int]:
        stack = []
        while head:
            stack.append(head.val)
            head = head.next
        stack.reverse()
        return stack



if __name__ == '__main__':
    node = head = ListNode(1)
    node.next = ListNode(3)
    node = node.next
    node.next = ListNode(2)
    print(Solution().reversePrint(head))
    print(Solution2().reversePrint(head))
